google/cloud/pubsub/samples/samples.cc

```diff
@@ -518,27 +518,33 @@ void Subscribe(google::cloud::pubsub::Subscriber subscriber,
   using google::cloud::future;
   using google::cloud::StatusOr;
   [](pubsub::Subscriber subscriber, pubsub::Subscription const& subscription) {
-    std::atomic<int> count{0};
-    google::cloud::promise<void> received_message;
-    auto result = subscriber.Subscribe(
+    std::mutex mu;
+    std::condition_variable cv;
+    bool done = false;
+    int message_count = 0;
+    auto session = subscriber.Subscribe(
         subscription, [&](pubsub::Message const& m, pubsub::AckHandler h) {
-          std::cout << "Received message " << m << "\n";
-          // Signal the waiting thread to detach the subscription after the
-          // first message.
-          if (count.load() == 0) received_message.set_value();
-          ++count;
-          // Only then ack the message, preventing any new requests from being
-          // issued.
+          std::cout << "Received message " << m << std::endl;
+          std::unique_lock<std::mutex> lk(mu);
+          ++message_count;
+          done = true;
+          lk.unlock();
+          cv.notify_one();
           std::move(h).ack();
         });
-    // Cancel the subscription as soon as the first message is received, and
-    // block until that happens.
-    received_message.get_future()
-        .then([&result](google::cloud::future<void>) { result.cancel(); })
-        .get();
+    // Wait until at least one message has been received.
+    std::unique_lock<std::mutex> lk(mu);
+    cv.wait(lk, [&done] { return done; });
+    lk.unlock();
+    // Cancel the subscription session.
+    session.cancel();
+    // Wait for the session to complete, no more callbacks can happen after this
+    // point.
+    std::cout << "GET=" << done << std::endl;
```

Comment:
Debugging?  It seems we know `done == true` here in any case.

Reply:
Removed.